adapters/dmx/dmx.go

```diff
@@ -0,0 +1,259 @@
+				H:      &inst.Banner.Format[0].H,
+				Format: inst.Banner.Format,
+			}
+
+			const intVal int8 = 1
+			var params dmxExt
+
+			source := (*json.RawMessage)(&inst.Ext)
+			if err := json.Unmarshal(*source, &params); err != nil {
+				errs = append(errs, err)
+			}
+			if params.Bidder.PublisherId == "" && params.Bidder.MemberId == "" {
+				var failedParams dmxParams
+				if err := json.Unmarshal(inst.Ext, &failedParams); err != nil {
+					errs = append(errs, err)
+					return nil, errs
+				}
+				imps = fetchAlternativeParams(failedParams, inst, ins, imps, banner, intVal)
+			} else {
+				imps = fetchParams(params, inst, ins, imps, banner, intVal)
+			}
+
+		}
+
+	}
+
+	dmxReq.Imp = imps
+
+	oJson, err := json.Marshal(request)
+
+	if err != nil {
+		errs = append(errs, err)
+	}
+	headers := http.Header{}
+	headers.Add("Content-Type", "Application/json;charset=utf-8")
+	reqBidder := &adapters.RequestData{
+		Method:  "POST",
+		Uri:     adapter.endpoint + addParams(sellerId), //adapter.endpoint,
+		Body:    oJson,
+		Headers: headers,
+	}
+
+	if dmxReq.User == nil {
+		if dmxReq.App == nil {
+			return nil, []error{errors.New("no user Id found and AppID, no request to DMX")}
+		}
+	}
+
+	reqsBidder = append(reqsBidder, reqBidder)
+	return
+}
+
+func (adapter *DmxAdapter) MakeBids(request *openrtb.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	var errs []error
+
+	if http.StatusNoContent == response.StatusCode {
+		return nil, nil
+	}
+
+	if http.StatusBadRequest == response.StatusCode {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected status code 400"),
+		}}
+	}
+
+	if http.StatusOK != response.StatusCode {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected response no status code"),
+		}}
+	}
+
+	var bidResp openrtb.BidResponse
+
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(5)
+
+	for _, sb := range bidResp.SeatBid {
+		for i := range sb.Bid {
+			bidType, err := getMediaTypeForImp(sb.Bid[i].ImpID, request.Imp)
+			if err != nil {
+				errs = append(errs, err)
+			} else {
+				b := &adapters.TypedBid{
+					Bid:     &sb.Bid[i],
+					BidType: bidType,
+				}
+				bidResponse.Bids = append(bidResponse.Bids, b)
+			}
+		}
+	}
+	return bidResponse, errs
+
+}
+
+func fetchAlternativeParams(params dmxParams, inst openrtb.Imp, ins openrtb.Imp, imps []openrtb.Imp, banner openrtb.Banner, intVal int8) []openrtb.Imp {
+	if params.TagId != "" {
+		ins = openrtb.Imp{
+			ID:     inst.ID,
+			TagID:  params.TagId,
+			Banner: &banner,
+			Ext:    inst.Ext,
+			Secure: &intVal,
+		}
+	}
+
+	if params.DmxId != "" {
+		ins = openrtb.Imp{
+			ID:     inst.ID,
+			TagID:  params.DmxId,
+			Banner: &banner,
+			Ext:    inst.Ext,
+			Secure: &intVal,
+		}
+	}
+	imps = append(imps, ins)
```

Comment:
What if `ins` is nil? It would be if both the `TagId` and `DmxId` are empty. An nil item in an `Imps` slice can have unpredictable behavior.